tests/test_read_data.py

```diff
@@ -0,0 +1,397 @@
+        backend=backend,
+        file_type=file_type,
+        files_per_partition=fpp,
+        add_filename=False,
+        input_meta=None,
+        columns=None,
+    )
+
+    # Compute the number of partitions in the resulting DataFrame
+    num_partitions = df.npartitions
+    # Assert that we have two partitions (since we have ~15KB total data and a blocksize of 10KB)
+    if fpp == 1:
+        assert (
+            num_partitions == NUM_FILES
+        ), f"Expected {NUM_FILES} partitions but got {num_partitions}"
+    elif fpp == NUM_FILES // 2:
+        assert (
+            num_partitions < NUM_FILES
+        ), f"Expected {NUM_FILES} partitions but got {num_partitions}"
+    elif fpp >= NUM_FILES:
+        assert num_partitions == 1, f"Expected 1 partition but got {num_partitions}"
+    else:
+        raise ValueError(f"Invalid fpp: {fpp}")
+    total_rows = len(df)
+    assert (
+        total_rows == NUM_FILES * NUM_RECORDS
+    ), f"Expected {NUM_FILES * NUM_RECORDS} rows but got {total_rows}"
+    if backend == "cudf":
+        import cudf
+
+        assert isinstance(df["id"].compute(), cudf.Series)
+    elif backend == "pandas":
+        assert isinstance(df["id"].compute(), pd.Series)
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        "pandas",
+        pytest.param("cudf", marks=pytest.mark.gpu),
+    ],
+)
+def test_read_data_blocksize_add_filename_jsonl(mock_multiple_jsonl_files, backend):
+    df = read_data_blocksize(
+        input_files=mock_multiple_jsonl_files,
+        backend=backend,
+        file_type="jsonl",
+        blocksize="128Mib",
+        add_filename=True,
+        input_meta=None,
+        columns=None,
+    )
+
+    assert "filename" in df.columns
+    file_names = df["filename"].unique().compute()
+    if backend == "cudf":
+        file_names = file_names.to_pandas()
+
+    assert len(file_names) == NUM_FILES
+    assert set(file_names.values) == {
+        f"test_{file_id}.jsonl" for file_id in range(NUM_FILES)
+    }
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        "pandas",
+        pytest.param("cudf", marks=pytest.mark.gpu),
+    ],
+)
+def test_read_data_blocksize_add_filename_parquet(mock_multiple_parquet_files, backend):
+    with pytest.raises(
+        ValueError,
+        match="add_filename and blocksize cannot be set at the same time for parquet files",
+    ):
+        read_data_blocksize(
+            input_files=mock_multiple_parquet_files,
+            backend=backend,
+            file_type="parquet",
+            blocksize="128Mib",
+            add_filename=True,
+            input_meta=None,
+            columns=None,
+        )
+
+
+@pytest.mark.parametrize(
+    "backend,file_type",
+    [
+        pytest.param("cudf", "jsonl", marks=pytest.mark.gpu),
+        pytest.param("cudf", "parquet", marks=pytest.mark.gpu),
+        ("pandas", "jsonl"),
+        pytest.param(
+            "pandas",
+            "parquet",
+            marks=pytest.mark.xfail(
```

Comment:
Marked it as `xfail` but this is something that we should look at. I would imagine that the errors in `test_read_data_select_columns` are related, but I could be wrong (or just tired looking at the same errors)

Reply:
This is demonstrating the note in the README:
"`filename` column cannot be selected when `backend="pandas"`, `filetype="parquet"` and `add_columns=True`"
?

This is saying that the "filename" column does not get added when `backend="pandas"` and `file_type="parquet"`? Or are you just saying it gets dropped by the `select_columns` function?

Reply:
`df.columns` doesn't return `filename` as a column, and I believe that's because `filename` is a reserved field in the pyarrow schema. However if we were to change our `filename` field to `file_name` / `path` things would work as expected. So if it sounds good I'll create a followup issue and do the renaming in the next PR (and also remove the xfail)